### ISSUE TITLE
Expand DockWidget API and add hookspec?

### ISSUE THREAD

**Issue by tlambert03** [napari/napari#1327]
## 🚀 Feature
In thinking about trying to integrate well-established community projects, like ilastik or deep lab cut, (or at least small modular parts of them)... we are, of course, a long way away from a full functional hook-based plugin architecture that would support all of their UX needs.

I'm wondering if we couldn't expand the dock widget API as a "loose" type of plugin that would immediately meet some needs, while we consider development of stricter hook-based plugins in parallel.  Specifically, when someone runs `viewer.window.add_dock_widget`, we pass back the `QtViewerDockWidget` object... and at the moment, it is a very thin wrapper a plain [`QDockWidget`](https://doc.qt.io/qt-5/qdockwidget.html).  However, it would be pretty easy to establish an API and write up some documentation that says "here's how you access the viewer that your widget is docked in"... and here's how you can use it, and perhaps inject some useful properties into their QWidget class.

We could define a new plugin hookspec that allowed plugins to declare these widgets, along with some metadata that would let us put a new item in the Plugins Menu, and let the end user select it and get that new dockwidget.

It's certainly not as "locked" down, and it does require more from developers (they'd need to pass a `QWidget`... or a function that we run `magicgui` on?), but I think we'd quickly get some very cool components (based on the many cool community dock widget examples we've seen on twitter and image.sc)

## Motivation

Letting advanced developers extend the napari gui with components that are well-beyond the current plugin design.

## Pitch

```python
# the spec

@napari_hook_specification
def napari_provide_dock_widget(): -> dict:
    """Return a dict that can be passed as kwargs to viewer.add_dock_widget"""

# the plugin

class MyWidget(QWidget):
    ...

    # these things would be true once added as a dock widget
    # we could maybe "inject" these properties for plugin developers too
    @property
    def dock_widget(self):
        return self.parent()

    @property
    def viewer(self):
        return self.get_dock_widget().parent()


@napari_hook_implementation
def napari_provide_dock_widget():
    return {
        'widget': MyWidget,
        'name': 'Cool Tracking Plugin',
        'area': 'right',
        'allowed_areas': ['right', 'left'],
        'shortcut': Alt+Shift-T,
    }

```

**Comment by GenevieveBuckley** [comment]
This is exactly what I want to do for some projects of mine, so I'm happy to see things moving in this direction! Can I try working out an initial implementation?

**Comment by jni** [comment]
@tlambert03 Love it! One modification I would suggest is:

```python
@napari_hook_specification
def napari_provide_dock_widget(): -> Tuple[QWidget, dict]:
    """Return a dict that can be passed as kwargs to viewer.add_dock_widget"""
```

To ensure that a widget is always provided.

**Comment by tlambert03** [comment]
> This is exactly what I want to do for some projects of mine, so I'm happy to see things moving in this direction! Can I try working out an initial implementation?

Sure that would be great!
The only non-obvious thing I might suggest is to implement it as a “historic” spec (and call it with call historic)... similar to how I wrote this one: https://github.com/napari/napari/issues/1272#issuecomment-633064212

**Comment by sofroniewn** [comment]
I have fears here about opening up the plugin API too quickly if we start accepting Qt widgets that can directly access the viewer is the way you describe - see comments from @royerloic here https://github.com/napari/napari/issues/1272#issuecomment-633697572 about insulation.

It basically means we have no control over backwards compatibility beyond the full API of napari (as opposed to the reader/ writers which are insulated).

On the other hand I see the immediate use cases, arguments that GUI's a different, and I can see an argument that a more flexible/ open plugin interface at this time is better than no plugin interface at this time, and then we can always add in more insulated ones

My big concern is that plugin developers will start to hate us and we'll get a reputation for instability if we keep breaking our APIs that affect plugins, and there are so many major refactors in the words right now (event handler, layer groups, world-coordinates, multi-canvas to name a few) that it seems hard to avoid.

I guess right now there just has to be some shared understanding that if you're adding a Qt element to a plugin you're a little more advanced, and we should make a space for the more advanced users.

Maybe even just making this `napari_provide_qt_dock_widget` to make it clear that this isn't going the magic gui route would make me feel better about this.

Final question - if a plugin wants to provide multiples doc widgets can this return a list?

**Comment by tlambert03** [comment]
I'm still not 100% sure I agree with the comments about insulation made in #1272.  I feel like there is a little bit of an illusion going on about exactly what we can and can't prevent users from doing in python (although I do like the point that "at least they know they're being naughty").  Moreover, this proposal does almost nothing that isn't there already.  By making something a dock widget, it already _can_ find the viewer (by just looking at the widget's parent).  So pretty much the only thing we're doing here is saying "hey, got a widget?  use this hook and if a user installs your plugin, we'll put it in the plugin menu for you".

it's the same as someone doing
```
pip install napari some-plugin

import napari, some_plugin

with napari.gui_qt():
    v = napari.Viewer()
    v.window.add_dock_widget(some_plugin.Widget, name="Plugin Widget")
```
(except with the ability to toggle it from the Plugins menu)

I am sensitive to the longevity argument for sure... but it's hard to even conceive a way in which we could eventually enable complicated GUI extensions in an insulated way such that developers aren't still just passing us a `QWidget`... so we might be waiting an extremely long time for that (without a professional team of full-time coders).  And the flipside of the stability argument is the idea that people _will_ figure out their own ways to do this, and there _will_ start to be as many different ways to do it as there are people trying to do it.   (remember, we can't control the use of the word plugin... and there are already github packages that call themselves "napari plugins" that don't use hooks... and each of them is going to have a different "install" procedure in their readme).   And then it might not be the plugin developers that hate us, but it could be the community that gets frustrated by heterogeneity...

**Comment by sofroniewn** [comment]
>  (remember, we can't control the use of the word plugin... and there are already github packages that call themselves "napari plugins" that don't use hooks... and each of them is going to have a different "install" procedure in their readme)

We can control though what get's installed into the bundled app through our "add plugin" dialog, and thinking 2 years from now that's how I see most biologists using napari + plugins, and it's that user experience that I think needs to be really robust and not a wild west (like so much of the tools we use)

> I feel like there is a little bit of an illusion going on about exactly what we can and can't prevent users from doing in python. Moreover, this proposal does almost nothing that isn't there already.

The distinction is really about what can a plugin that ultimately gets installed via our "add plugin" dialog can do, not about what is possible more broadly. For people that build their own "plugins" that use napari as a component of half and half etc that people install themselves we have no control over, and so in many ways for them we don't really need a plugin interface/ hook specs, they can just call `add_dock_widget` themselves.

Now I do appreciate the complexity of GUI widgets, so having a hook spec with `qt` in the name seems the right middle ground for where we are now, and if we ever migrate away from that, deprecate that it will be very clear why - because we moved to an API that didn't allow passage of a `qt` object

**Comment by tlambert03** [comment]
yeah, I don't want wild west either.  So definitely happy to keep this conversation going and get more input from all @napari/core-devs 

but while we're talking about what a plugin can and can't do that get's discovered via some hookspec, don't forget that we are importing their module to find their hook implementations.  And once we run `importlib.import_module("some_plugin")`, i hate to say it, but they can pretty much do or monkeypatch whatever they want!  Again though, I do appreciate @royerloic's point that "at least they know they're being naughty" ... and that's super important.  But there's a big spectrum of just _how_ naughty they feel like they're being.  (and the less we provide for them, the more likely they are to feel justified in being naughty)

**Comment by 0x00b1** [comment]
> My big concern is that plugin developers will start to hate us and we'll get a reputation for instability if we keep breaking our APIs that affect plugins, and there are so many major refactors in the words right now (event handler, layer groups, world-coordinates, multi-canvas to name a few) that it seems hard to avoid.

Totally. 

In this type of conversation I always return to the @torvalds commitment for Linux, “don’t break user space.” It’s especially wise for scientific software where code should be able to run, without fuss, decades from now. Furthermore, it has the benefit of simplify decision making by weighing the benefits of a proposed feature against the possible complications of maintaining the “don’t break user space” commitment, e.g., in this spirit, does this PR make it harder to transition to a new version of Qt? A different widget framework? A different execution environment?

**Comment by GenevieveBuckley** [comment]
Those are all good points @sofroniewn 

Can you clarify what you mean by "going the magicgui route"?

**Comment by sofroniewn** [comment]
> Can you clarify what you mean by "going the magicgui route"?

This would mean that we accept python functions with typed signatures and then internally we use magicgui (or some similar process) to generate the qt elements.

Note here that plugin developers are not using magicgui and do not depend on it, napari is using magicgui, and so when the qt5 -> qt6 release happens ([which will be in 2020!!!](https://www.phoronix.com/scan.php?page=news_item&px=Qt-6.0-Release-Date#:~:text=The%20plan%20at%20this%20point,0%20released%20on%201%20December.)) we just have to update magicgui and all our plugins keep on working. Similarly in 2 years if we swap to a web-based UI that calls python code we just make a new version of magicgui that makes web-based UI elements and everything keeps on working.

**Comment by GenevieveBuckley** [comment]
> 
> 
> > Can you clarify what you mean by "going the magicgui route"?
> 
> This would mean that we accept python functions with typed signatures and then internally we use magicgui (or some similar process) to generate the qt elements.
> 
> Note here that plugin developers are not using magicgui and do not depend on it, napari is using magicgui, and so when the qt5 -> qt6 release happens ([which will be in 2020!!!](https://www.phoronix.com/scan.php?page=news_item&px=Qt-6.0-Release-Date#:~:text=The%20plan%20at%20this%20point,0%20released%20on%201%20December.)) we just have to update magicgui and all our plugins keep on working. Similarly in 2 years if we swap to a web-based UI that calls python code we just make a new version of magicgui that makes web-based UI elements and everything keeps on working.

So you think the magicgui route is good or bad? 

(Your first comment made it sound like you thought that would be bad, but this comment sounds like you think it's a good idea)

**Comment by sofroniewn** [comment]
> So you think the magicgui route is good or bad?

All of this is more nuanced than good or bad for me, I really can see pros/ cons to all the approaches, but I probably prefer a magicgui-like approach where Python functions/ objects are passed instead of full `Qt` objects, but I'd want to discuss more the current method of passing layers directly to magicgui before solidifying that in the plugin architecture.

**Comment by tlambert03** [comment]
yeah, I see these as two independent things, which both have their merits and which both may be implemented at some point.  constructing guis for people (using, e.g. magicgui) definitely has the advantage that we can swap out the backend anytime, and I do see that as a longer-term goal.  For the moment, actually connecting signals/interactivity would be rather hard with magicgui if _we_ do the wrapping for them... (all of the magicgui examples connect events after the widget is created).  If we accepted a class, then they could declare those connections.  So maybe someday a magicgui-like solution will be great here.

this is all why I stated in my first post:

> expand the dock widget API as a "loose" type of plugin that would immediately meet some needs, while we consider development of stricter hook-based plugins in parallel.

If we want to abstract all of the interactivity that someone could achieve by passing us a QWidget (instead of us building it for them), then we're basically rewriting Qt ourselves.  It's absolutely a "desirable" goal... I just don't see it as achievable.  And maybe that's fine, maybe we just don't really support rich extension of the gui like that.  

If it would help, i would (literally) not be opposed to this hookspec being called:
```python
@napari_hook_specification
def napari_provide_qt_widget_that_is_not_guaranteed_to_be_supported_in_the_future():
```

**Comment by 0x00b1** [comment]
@tlambert03 A common approach in cross-platform applications that need to stay agnostic about their underlying widget toolkit is to shim (and use) just HTML widgets, e.g. input and select. It provides a simple common denominator that is more than likely going to match semantics with whatever widget toolkit you’d consider using.

**Comment by GenevieveBuckley** [comment]
So I think what I'm most concerned about is what you might call the lowest common denominator use case for plugins. That is something that can:
* Install into the bundled napari app
* Add itself to the plugin menu list, so users can click to start the plugin
* When the plugin is launched it should add a little GUI, so the user can select the input layer(s) plus any other needed parameters
* Allow the user to be able to click a button to run it, and spit back (or save) a layer with results

I'd be pretty happy if all of this was possible in a very structured/limited way using hookspecs. 
But so far I don't think we have a hookspec to take in layers, compute something, and return a result as a layer; and we definitely don't have a way for users to select parameters from a plugin GUI.

We do have writing our own scripts with magicgui that then call napari, but the conda setup is still tricky for users. But fairly frequently I have conversations with my colleagues where someone is stuck because they're not super familiar with conda and can't quite get the setup working. The two biggest things that would improve this are (1) simplifying installation with app bundles, and (2) giving them a plugin GUI they can click.

Do we have a way to move forward on this type of use case, or is everything on hold until we work out the overall architecture?

**Comment by sofroniewn** [comment]
> Do we have a way to move forward on this type of use case, or is everything on hold until we work out the overall architecture?

I would hold on this for a little, it will certainly need some dedicated discussion.

**Comment by quantumjot** [comment]
A few thoughts from someone who's trying to build an extension for napari.  One of the things that has been so useful with napari, is that we can easily customize it directly for different tasks by bolting on extra pieces of code that augment the functionality. I really believe that's one of the true strengths here - so features like the `add_dock_widget` have been invaluable.

However, i'm totally convinced by the 'dependency hell' argument against such extensions being installed directly through napari - it seems that is going to be the biggest challenge of all. Some insulation/containerization could help, but my experience is that those come with performance overheads, especially when dealing with large datasets.  

Perhaps you could have something akin to a napari extension zoo, where extensions (not plugins) live and are discoverable.  These extensions would make use of an API similar to the one Talley describes, but with the caveat of needing certain environment to work (not dissimilar to the kernels idea in jupyter notebooks).

BTW - I removed the `napari-plugin` tag for arboretum in light of these discussions, since we're not using any of the existing hookspecs.

**Comment by neuromusic** [comment]
> Perhaps you could have something akin to a napari extension zoo, where extensions (not plugins) live and are discoverable.

I think this makes a lot of sense to me. You see similar concepts in a lot of different platforms that support 3rd-party tooling... On Github, for example, you have Apps (which have a very wide surface, but are siloed from one another) vs Actions (which are more restrictive, but offer compatibility and composability). Lots of platforms separate Themes from Extensions from Recipes.

So yeah, having a distinct concept like an "Extension" which is more restrictive in what the extension developer can do, but offers a stronger guarantee of compatibility with other Extensions could be highly valuable for plugin/extension developers and end users.

In napari-land, I could imagine things like Tasks (take in data & parameters, and return data) or Components (specialized viz or UI) or Apps (a bundle of Tasks & Components which work together) or Workflows (a DAG of Tasks). They would have different areas on the Zoo or App Store where users find them. On the implementation-side, they could all use the "plugin" API, but would be defined in part by which part of the plugin API they are restricted to touching (perhaps a restricted set of allowed dependencies, too). 

And at the end of the day, an enterprising developer could still build a monolithic plugin, incompatible with the rest of the ecosystem by ignoring those conventions and working directly with the plugin API and telling their users to install via pip.

**Comment by neuromusic** [comment]
Oh wait, I think I inverted @quantumjot 's Extension/Plugin relationship. Extension = less restrictive API surface, more restrictive allowed dependencies. Still, the broader concepts of different flavors of "plugins" with different rules could be super super valuable.

**Comment by tlambert03** [comment]
Ok, sounds like wording is very important here, and everyone is a bit more comfortable opening up various APIs as long as we protect the word "plugin".  I can certainly get behind that.  I do think we need _some_ sort of avenue to promote the amazing work that early adopters are doing, and I genuinely do think we will kill some of the buzz/excitement if we make everyone wait until we come up with some abstraction between the viewer API and some plugin API (that may or may not ever come). so how about we "formalize" the concept of a napari extension:

**Proposal for napari "extensions":**
- extensions are marked with a `@napari_extension` decorator imported from `napari-plugin-engine` (behind the scenes they will be registered using `napari-plugin-engine` PluginManager ... but that needn't have any implications beyond the fact that it would be silly to replicate all that machinery)
- (they could be labeled with `napari-extension` in github)
- Extension hooks WILL be passed napari objects, like `Viewer` and `Layer`, but will NOT be hooked into the public napari "API" (in other words, an extension would never modify the behavior of a napari function like `Viewer.open`, whereas a plugin might).
- We are unlikely to "use" data returned from an extension, instead, they are expected to call the native APIs of the napari objects they are given.
- we make it clear that it is up to the extension author to update their extensions to maintain compatibility with the napari API.  We make no guarantees that changes to napari will not break extensions.
- we can provide at least some way for an extension to declare what version of napari it was designed for... but no runtime checking will be performed.

I think the hookspec proposed in my first post here would be a reasonable place to start with extensions.  They provide a widget, and some metadata, and we will make it visible in some "extensions" menu (we can create a seperate menu for extensions or a sub-menu of plugins, doesn't matter to me) that a user can find.  this would already be great for @quantumjot's stuff, the [`napari-omero`](https://github.com/tlambert03/napari-omero), etc...

thoughts?

**Comment by sofroniewn** [comment]
Hi @tlambert03 thanks for keeping this thread going. Following on, here are a couple additional thoughts on how people could maybe build on top of napari, thinking more from the user/ builder perspective rather than implementation perspective. Note that what follows are just some proposals and ideas, not anything written in stone:

### Extensions
I see extensions as interactivity focused. They allow people to customize the interactivity napari provides, provide their own GUI elements. You wouldn't use an extension "headless" because there is no need for custom plots, mouse clicks, or key presses "headless". Extensions are additive, and you're generally only using one or two at a time and explicitly turning them on or off when you want to use them.

I see extensions as being able to do four main things:
- provide custom dock widgets (with say custom plots or data displays)
- provide custom mouse functions
- provide custom keybindings
- connect into our event system to provide callbacks

You might have 50 extensions installed, but only 2 enabled A and B. Extension A might provide 3 dock widgets and 2 mouse functions, and 3 callbacks. Extension B might provide 2 dock widgets, 1 mouse function, and 3 keybindings, and no callbacks. These would be combined additively so that you'd have access to 5 dock widgets, 3 mouse functions, 3 keybindings, and 3 callbacks.

Examples of extensions could now be
- `napari-console` if we pull the console out of the main repo
- `napari-movie` see https://github.com/guiwitz/naparimovie and by #780 @guiwitz
- https://github.com/hci-unihd/covid-if-annotations by @constantinpape
- https://github.com/quantumjot/arboretum by @quantumjot
- https://github.com/SainsburyWellcomeCentre/cellfinder by @adamltyson
- https://github.com/4DNucleome/PartSeg by @Czaki (I know we still need to have multicanvas support first)
- https://github.com/tlambert03/napari-omero

Note here would these extensions be able to provide an actual Qt widget - what if it is PySide2 vs PyQt5, what happens during the Qt6 transition in December? Are all these extensions provided the full `viewer` object? Will that lead to extensions breaking when they are used together as they will be trying to do incompatible things. Can we pass only the `ViewerModel` to our extensions, not the full `Viewer` (but that might not allow functionality like screenshot). If we simplify our models like in #1353, might be only provide subsets of the models to the extensions?

Ideally extensions would be able to customize stylings and theme, but that conflicts somewhat with the "additive" nature of extensions, now we'd need a priority status.

### Plugins
I see plugins as IO/ analysis focused. These should be pure python and highly independent of any GUI elements. They could often be described as python libraries. They probably need to understand `napari-types` like images, labels, shapes etc. but not napari specific objects. They should be modular, composable, and easily runable "headless" as they are pure python.

I see the two main types of plugins as doing
- IO, taking a path and providing a list of layer type data or vice versa, taking a list of layer type data and saving it to a path
- Analysis, taking a list of layer type data operating on it and returning a list of layer type data.

You might have 50 plugins installed each of which may provide many of those IO or analysis functions that are callable from the GUI (either from drop downs, or very simple autogenerate GUI elements, which allow additional parameter values to be passed). Note here all drop downs are autogenerated, all GUI elements are autogenerated and focused around parameter input (and not say custom plotting, which would appear in an extension).

Examples of plugins could now be
- scikit-image
- https://github.com/AllenCellModeling/napari-aicsimageio by @JacksonMaxfield
- https://github.com/manzt/napari-lazy-openslide by @manzt 
- https://github.com/tlambert03/napari-dv
- https://github.com/transformify-plugins/segmentify

Note here these plugins are functional and stateless so no napari objects like layers or the viewer would pass through them, only the associated data.

### Workflows
I see workflows as simple python scripts that describe the chaining together of IO and analysis calls (which use plugins as needed). They should be pure python, runnable headless, and the CLI tool to run them should just be python. They should be recordable from the GUI (like a "macro recorder", like ImageJ) or maybe one day they can be made via a graphical programming language (like "protocols" in Icy), but they should always be valid runnable python code. These would allow a user to a workflow composed of many plugins, with say registration, segmentation, and tracking steps.

----------------------------------------------------------------------------------------

### How does it all fit together?

I can see some problems where someone might be building a tool that wants to provide both extension-like and plugin-like functionality, and maybe also some workflows. Should that be allowed in the same repo. Probably from a convenience standpoint, yes, but I'd prefer a world where not all our plugins lived in repos that also had GUI code/ pyqt dependencies etc as the goal is to make install of that super robust everywhere.

I'm still a little nervous about letting an extension return a Qt widget that has the full viewer on it, as I worry that in a couple years we'll have lots of extensions that no longer work as Qt has changed or our viewer has changed, but I do see the value in opening up this approach now as extensions will be installable into the bundled app and I think for that reason are preferable to people using napari as a component in their own tool which then must be installed and run completely independently.

**Comment by Czaki** [comment]
I'm also interested in add plugin system for widget. Maybe good idea to be use `Protocols` and `TypedDict`? So:

From my experience I prefer to use `Protocols` with `classmethod` than description in additional structure. `runtime_checkable` decorator allow to check which protocol is implemented. 

Additionally napari does not use all popular widgets, and unused one are styled symilary to plain `QWidget`. There should be some instruction how to fix it and how prepare style matching others and some hook for "Toggle Theme" 

And maybe some instruction to create additional icons which will match napari style (you spend big effort to get clean and nice view of napari window). 

And build more complex plugins need better API stability.
